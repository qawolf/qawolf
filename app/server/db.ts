import cuid from "cuid";
import knex from "knex";
import path from "path";
import { Client } from "pg";

import environment from "./environment";

const TEST_DATABASE_NAME = `test_${cuid()}`;

export const createTestDb = async (): Promise<void> => {
  const client = new Client();
  await client.connect();
  await client.query(`CREATE DATABASE ${TEST_DATABASE_NAME}`);
  await client.end();
};

let databaseReadyPromise = Promise.resolve();
if (environment.NODE_ENV === "test") {
  databaseReadyPromise = createTestDb();
}

export const db = knex({
  client: "pg",
  connection: async () => {
    await databaseReadyPromise;

    if (environment.NODE_ENV === "test") {
      return {
        database: TEST_DATABASE_NAME,
        password: process.env.PGPASSWORD,
        user: process.env.PGUSER,
      };
    }

    const connection: knex.PgConnectionConfig = {
      connectionString: environment.DATABASE_URL,
    };
    if (environment.DATABASE_SSL) connection.ssl = environment.DATABASE_SSL;

    return connection;
  },
});

export const dropTestDb = async (): Promise<void> => {
  await db.destroy();
  const client = new Client();
  await client.connect();
  await client.query(`DROP DATABASE ${TEST_DATABASE_NAME}`);
  await client.end();
};

export const migrateDb = async (): Promise<void> => {
  await db.migrate.latest({ directory: path.join(__dirname, "migrations") });
};
