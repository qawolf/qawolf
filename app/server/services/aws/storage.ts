import { PutObjectCommand, S3 } from "@aws-sdk/client-s3";
import { S3RequestPresigner } from "@aws-sdk/s3-request-presigner";
import { createRequest } from "@aws-sdk/util-create-request";
import { formatUrl } from "@aws-sdk/util-format-url";

import environment from "../../environment";
import { SaveArtifacts } from "../../types";

type GetArtifactsOptions = {
  excludeVideo?: boolean;
  name: string;
};

const client = new S3({
  credentials: {
    accessKeyId: environment.QAWOLF_AWS_ACCESS_KEY_ID,
    secretAccessKey: environment.QAWOLF_AWS_SECRET_ACCESS_KEY,
  },
  region: environment.QAWOLF_AWS_REGION,
});

const expiresInSeconds = 35 * 60;

export const createStorageReadAccessUrl = (fileName: string): string => {
  return `https://${environment.QAWOLF_AWS_S3_BUCKET}.s3.${environment.QAWOLF_AWS_REGION}.amazonaws.com/${fileName}`;
};

// https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/s3-example-creating-buckets.html#s3-create-presigendurl
export const createStorageWriteAccessUrl = async (
  fileName: string
): Promise<string> => {
  const signer = new S3RequestPresigner({ ...client.config });

  const request = await createRequest(
    client,
    new PutObjectCommand({
      Bucket: environment.QAWOLF_AWS_S3_BUCKET,
      Key: fileName,
    })
  );

  const signedRequest = await signer.presign(request, {
    expiresIn: expiresInSeconds,
  });

  const url = formatUrl(signedRequest);

  return url;
};

export const getArtifactsOptions = async ({
  excludeVideo,
  name,
}: GetArtifactsOptions): Promise<SaveArtifacts> => {
  const gifFileName = `${name}.gif`;
  const logsFileName = `${name}.txt`;
  const videoFileName = `${name}.mp4`;

  return {
    gifUrl: excludeVideo
      ? null
      : await createStorageWriteAccessUrl(gifFileName),
    logsUrl: await createStorageWriteAccessUrl(logsFileName),
    videoUrl: excludeVideo
      ? null
      : await createStorageWriteAccessUrl(videoFileName),
  };
};
