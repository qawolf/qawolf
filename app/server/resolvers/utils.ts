/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { Transaction } from "knex";

import { AuthenticationError } from "../errors";
import { Logger } from "../Logger";
import { findGroup } from "../models/group";
import { findSuite } from "../models/suite";
import { findTest } from "../models/test";
import { Team, Test, User } from "../types";

type EnsureGroupAccess = {
  group_id: string;
  logger: Logger;
  teams: Team[] | null;
  trx?: Transaction;
};

type EnsureSuiteAccess = {
  logger: Logger;
  suite_id: string;
  teams: Team[] | null;
};

type EnsureTeams = {
  logger: Logger;
  teams: Team[] | null;
};

type EnsureTeamAccess = {
  logger: Logger;
  team_id: string;
  teams: Team[] | null;
};

type EnsureTestAccess = {
  logger: Logger;
  teams: Team[] | null;
  test?: Test;
  test_id?: string;
};

type EnsureUser = {
  logger: Logger;
  user: User | null;
};

/**
 * @summary Throws if `teams` is falsy
 * @returns `teams`
 */
export const ensureTeams = ({ logger, teams }: EnsureTeams): Team[] => {
  const log = logger.prefix("ensureTeams");

  if (!teams) {
    log.error("teams not provided");
    throw new AuthenticationError("no teams");
  }

  return teams;
};

export const ensureGroupAccess = async ({
  group_id,
  logger,
  teams,
  trx,
}: EnsureGroupAccess): Promise<Team> => {
  const log = logger.prefix("ensureGroupAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);

  log.debug("ensure teams", teamIds, "can access group", group_id);

  const group = await findGroup(group_id, { logger, trx });
  const selectedTeam = teams!.find((team) => group.team_id === team.id);

  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access group", group_id);
    throw new AuthenticationError("cannot access group");
  }

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `suite.team_id`. Can also
 *   throw if `suite_id` doesn't match any suite.
 * @returns The team from `teams` that matches `suite.team_id`, if we didn't throw.
 */
export const ensureSuiteAccess = async ({
  logger,
  teams,
  suite_id,
}: EnsureSuiteAccess): Promise<void> => {
  const log = logger.prefix("ensureSuiteAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  log.debug("ensure teams", teamIds, "can access suite", suite_id);

  const suite = await findSuite(suite_id, { logger });

  if (!teamIds.includes(suite.team_id)) {
    log.error("teams", teamIds, "cannot access suite", suite_id);
    throw new Error("Teams cannot access suite");
  }
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `team_id`
 * @returns The team from `teams` that matches `team_id`, if we didn't throw.
 */
export const ensureTeamAccess = ({
  logger,
  team_id,
  teams,
}: EnsureTeamAccess): Team => {
  const log = logger.prefix("ensureTeamAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  const selectedTeam = teams!.find((team) => team.id === team_id);
  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access team", team_id);
    throw new AuthenticationError("cannot access team");
  }

  log.debug("ensured teams", teamIds, "can access team", team_id);

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `test.team_id`. Can also
 *   throw if `test_id` doesn't match any test.
 * @returns The team from `teams` that matches `test.team_id`, if we didn't throw.
 */
export const ensureTestAccess = async ({
  logger,
  teams,
  test,
  test_id,
}: EnsureTestAccess): Promise<Team> => {
  const log = logger.prefix("ensureTestAccess");

  const finalTeams = ensureTeams({ logger, teams });
  const teamIds = finalTeams.map((t) => t.id);

  log.debug("ensure teams", teamIds, "can access test", test?.id || test_id);

  const finalTest = test || (await findTest(test_id, { logger }));
  const team = finalTeams.find((t) => finalTest.team_id === t.id);

  if (!team) {
    log.error("teams", teamIds, "cannot access test", test?.id || test_id);
    throw new AuthenticationError("cannot access test");
  }

  return team;
};

/**
 * @summary Throws if `user` is falsy
 * @returns `user` if we didn't throw
 */
export const ensureUser = ({ logger, user }: EnsureUser): User => {
  const log = logger.prefix("ensureUser");

  if (!user) {
    log.error("user not provided");
    throw new AuthenticationError("no user");
  }

  return user;
};
