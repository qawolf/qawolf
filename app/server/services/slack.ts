import axios from "axios";
import querystring from "querystring";

import environment from "../environment";
import { Logger } from "../Logger";
import { SlackWebhook } from "../types";
import { cuid } from "../utils";

type FindSlackWebhook = {
  logger: Logger;
  redirect_uri: string;
  slack_code: string;
};

type SlackWebhookData = {
  data: {
    incoming_webhook: {
      channel: string;
      configuration_url: string;
      url: string;
    };
    team: { name: string };
  };
};

const SLACK_ACCESS_URL = "https://slack.com/api/oauth.v2.access";
const SLACK_AUTHORIZE_URL = "https://slack.com/oauth/v2/authorize";

const buildSlackWebhook = (data: SlackWebhookData["data"]): SlackWebhook => {
  return {
    settings_url: data.incoming_webhook.configuration_url,
    slack_channel: data.incoming_webhook.channel,
    slack_team: data.team.name,
    webhook_url: data.incoming_webhook.url,
  };
};

// Slack documentation
// https://api.slack.com/legacy/oauth#authenticating-users-with-oauth__the-oauth-flow__step-1---sending-users-to-authorize-andor-install
export const createSlackIntegrationUrl = (redirect_uri: string): string => {
  const queryString = querystring.stringify({
    client_id: environment.SLACK_CLIENT_ID,
    redirect_uri: new URL(redirect_uri, environment.APP_URL).href,
    scope: "incoming-webhook",
    state: cuid(),
  });

  return `${SLACK_AUTHORIZE_URL}?${queryString}`;
};

// Slack documentation https://api.slack.com/methods/oauth.v2.access
export const findSlackWebhook = async ({
  logger,
  redirect_uri,
  slack_code,
}: FindSlackWebhook): Promise<SlackWebhook> => {
  const log = logger.prefix("findSlackWebhook");

  try {
    const { data }: SlackWebhookData = await axios.post(
      SLACK_ACCESS_URL,
      querystring.stringify({
        code: slack_code,
        redirect_uri: new URL(redirect_uri, environment.APP_URL).href,
      }),
      {
        auth: {
          password: environment.SLACK_CLIENT_SECRET,
          username: environment.SLACK_CLIENT_ID,
        },
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
        },
      }
    );

    log.debug(
      `webhook channel ${data.incoming_webhook.channel} and team ${data.team.name}`
    );

    return buildSlackWebhook(data);
  } catch (error) {
    log.error("error", error);
    throw new Error("No webhook returned by Slack");
  }
};
