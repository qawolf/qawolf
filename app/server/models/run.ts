import { db } from "../db";
import { sendAlert } from "../services/alert/send";
import {
  createStorageReadAccessUrl,
  getArtifactsOptions,
} from "../services/aws/storage";
import { updateCommitStatus } from "../services/gitHub/app";
import {
  ModelOptions,
  Run,
  RunnerRun,
  RunResult,
  RunStatus,
  SuiteRun,
  Test,
} from "../types";
import { cuid, minutesFromNow } from "../utils";
import { decrypt } from "./encrypt";
import { buildEnvironmentVariablesForGroup } from "./environment_variable";

type CreateRunsForTests = {
  suite_id: string;
  tests: Test[];
};

type FindLatestRuns = {
  group_id: string;
  test_id: string;
};

export type UpdateRun = {
  code?: string;
  current_line?: number;
  id: string;
  started_at?: string;
  status?: RunStatus;
};

export const countPendingRuns = async ({
  logger,
  trx,
}: ModelOptions): Promise<number> => {
  const log = logger.prefix("countPendingRuns");

  const result = await (trx || db)("runs")
    .count("*", { as: "count" })
    .from("runs")
    .leftJoin("runners", "runners.run_id", "runs.id")
    .where({ "runners.id": null, started_at: null })
    .first();

  log.debug(result);

  return Number(result.count);
};

export const createRunsForTests = async (
  { suite_id, tests }: CreateRunsForTests,
  { logger, trx }: ModelOptions
): Promise<Run[]> => {
  const log = logger.prefix("createRunsForTests");

  log.debug("suite", suite_id);

  const created_at = minutesFromNow();

  const runs = tests.map(({ code, id: test_id }) => ({
    code,
    created_at,
    current_line: null,
    id: cuid(),
    status: "created" as RunStatus,
    suite_id,
    test_id,
  }));
  await (trx || db)("runs").insert(runs);

  log.debug(
    "created",
    runs.map((r) => r.id)
  );

  return runs;
};

export const findLatestRuns = async (
  { group_id, test_id }: FindLatestRuns,
  { logger, trx }: ModelOptions
): Promise<SuiteRun[]> => {
  const log = logger.prefix("findLatestRuns");

  log.debug("group", group_id, "test", test_id);

  const runs = await (trx || db)
    .select("runs.*" as "*")
    .select("tests.name AS test_name")
    .select("tests.deleted_at AS test_deleted_at")
    .from("runs")
    .innerJoin("suites", "runs.suite_id", "suites.id")
    .innerJoin("tests", "runs.test_id", "tests.id")
    .where({ group_id, test_id })
    .orderBy("created_at", "desc")
    .limit(10);

  log.debug(`found ${runs.length} runs`);

  return runs.map((run: Run & { test_deleted_at: string | null }) => {
    const gif_url = run.completed_at
      ? createStorageReadAccessUrl(`${run.id}.gif`)
      : null;

    return { ...run, gif_url, is_test_deleted: !!run.test_deleted_at };
  });
};

export const findPendingRun = async (
  { needs_runner }: { needs_runner?: true },
  { trx }: ModelOptions
): Promise<Pick<Run, "created_at" | "id"> | null> => {
  let query = (trx || db)
    .select("runs.id" as "*")
    .select("runs.created_at" as "*")
    .from("runs")
    .where({ completed_at: null });

  if (needs_runner) {
    query = query
      .leftJoin("runners", "runners.run_id", "runs.id")
      .where({ "runners.id": null, started_at: null });
  }

  const result = await query.orderBy("runs.created_at", "asc").first();

  return result || null;
};

export const findRun = async (
  id: string,
  { trx }: ModelOptions
): Promise<Run> => {
  const run = await (trx || db).select("*").from("runs").where({ id }).first();
  if (!run) throw new Error("run not found " + id);
  return run;
};

export const findRunResult = async (
  id: string,
  { logger, trx }: ModelOptions
): Promise<RunResult> => {
  const run = await findRun(id, { logger, trx });

  let logs_url: string | null = null;
  let video_url: string | null = null;

  if (run.completed_at) {
    logs_url = createStorageReadAccessUrl(`${run.id}.txt`);
    video_url = createStorageReadAccessUrl(`${run.id}.mp4`);
  }

  return { ...run, logs_url, video_url };
};

export const findRunsForSuite = async (
  suite_id: string,
  { trx }: ModelOptions
): Promise<SuiteRun[]> => {
  const runs = await (trx || db)
    .select("runs.*" as "*")
    .select("tests.deleted_at AS test_deleted_at")
    .select("tests.name AS test_name")
    .from("runs")
    .innerJoin("tests", "runs.test_id", "tests.id")
    .where({ "runs.suite_id": suite_id })
    .orderBy("test_name", "asc");

  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  return runs.map((run: any) => {
    const gif_url = run.completed_at
      ? createStorageReadAccessUrl(`${run.id}.gif`)
      : null;

    return {
      ...run,
      gif_url,
      is_test_deleted: !!run.test_deleted_at,
    };
  });
};

export const findSuiteRunForRunner = async (
  run_id: string,
  { logger, trx }: ModelOptions
): Promise<RunnerRun | null> => {
  return (trx || db).transaction(async (trx) => {
    const row = await trx
      .select("runs.*")
      .select("suites.environment_variables AS environment_variables")
      .select("suites.group_id AS group_id")
      .select("suites.team_id AS team_id")
      .select("tests.version AS test_version")
      .from("runs")
      .innerJoin("suites", "runs.suite_id", "suites.id")
      .innerJoin("tests", "runs.test_id", "tests.id")
      .andWhere({ "runs.id": run_id })
      .first();

    if (!row) return null;

    const custom_variables = row.environment_variables
      ? JSON.parse(decrypt(row.environment_variables))
      : {};
    const env = await buildEnvironmentVariablesForGroup(
      { custom_variables, group_id: row.group_id, team_id: row.team_id },
      { logger, trx }
    );

    return {
      artifacts: await getArtifactsOptions({ name: row.id }),
      code: row.code,
      env,
      id: row.id,
      test_id: row.test_id,
      version: row.test_version,
    };
  });
};

export const updateRun = async (
  { id, ...options }: UpdateRun,
  { logger, trx }: ModelOptions
): Promise<Run> => {
  const log = logger.prefix("updateRun");

  const timestamp = minutesFromNow();

  const updates: Partial<Run> = {
    ...options,
    updated_at: timestamp,
  };

  if (options.status === "fail" || options.status === "pass") {
    updates.completed_at = timestamp;
  }

  const run = await findRun(id, { logger, trx });
  await (trx || db)("runs").where({ id }).update(updates);

  if (updates.completed_at && run.suite_id) {
    await Promise.all([
      sendAlert({ logger, suite_id: run.suite_id }),
      updateCommitStatus({ logger, suite_id: run.suite_id }),
    ]);
  }

  log.debug("updated", id, updates);

  return { ...run, ...updates };
};
