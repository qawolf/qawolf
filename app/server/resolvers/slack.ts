import { db } from "../db";
import { updateGroup } from "../models/group";
import { createIntegration } from "../models/integration";
import { createSlackIntegrationUrl, findSlackWebhook } from "../services/slack";
import {
  Context,
  CreateSlackIntegrationMutation,
  CreateUrlMutation,
  Integration,
} from "../types";
import { ensureGroupAccess, ensureUser } from "./utils";

export const createSlackIntegrationResolver = async (
  _: Record<string, unknown>,
  { group_id, redirect_uri, slack_code }: CreateSlackIntegrationMutation,
  { logger, teams }: Context
): Promise<Integration> => {
  const log = logger.prefix("createSlackIntegrationResolver");
  const team = await ensureGroupAccess({ group_id, logger, teams });

  const slackWebhook = await findSlackWebhook({
    logger,
    redirect_uri,
    slack_code,
  });

  return db.transaction(async (trx) => {
    const integration = await createIntegration(
      {
        settings_url: slackWebhook.settings_url,
        slack_channel: slackWebhook.slack_channel,
        team_id: team.id,
        team_name: slackWebhook.slack_team,
        type: "slack",
        webhook_url: slackWebhook.webhook_url,
      },
      { logger, trx }
    );
    log.debug("created integration", integration.id);

    await updateGroup(
      {
        id: group_id,
        is_email_enabled: false,
        alert_integration_id: integration.id,
      },
      { logger, trx }
    );
    log.debug("updated group", group_id);

    return integration;
  });
};

export const createSlackIntegrationUrlResolver = (
  _: Record<string, unknown>,
  { redirect_uri }: CreateUrlMutation,
  { logger, user: contextUser }: Context
): string => {
  const log = logger.prefix("createSlackIntegrationUrlResolver");
  const user = ensureUser({ logger, user: contextUser });

  const url = createSlackIntegrationUrl(redirect_uri);
  log.debug("url for user", user.id);

  return url;
};
