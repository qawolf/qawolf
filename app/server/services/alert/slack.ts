import { IncomingWebhook, IncomingWebhookSendArguments } from "@slack/webhook";

import environment from "../../environment";
import { Logger } from "../../Logger";
import { findIntegration } from "../../models/integration";
import { findUsersForTeam } from "../../models/user";
import { Group, Suite, SuiteRun, User } from "../../types";
import { randomChoice } from "../../utils";

type BuildSuiteMessage = {
  group: Group;
  runs: SuiteRun[];
  suite: Suite;
  user: User | null;
};

type PostMessageToSlack = {
  message: IncomingWebhookSendArguments;
  webhook_url: string;
};

type SendSlackAlert = {
  group: Group;
  logger: Logger;
  runs: SuiteRun[];
  suite: Suite;
};

export const buildMessageForSuite = ({
  group,
  runs,
  suite,
  user,
}: BuildSuiteMessage): IncomingWebhookSendArguments => {
  const wolfName = user?.wolf_name || "Spirit";
  const wolfVariant = user?.wolf_variant || "white";

  const failingRuns = runs.filter((r) => r.status === "fail");

  const suiteHref = new URL(`/tests/${suite.id}`, environment.APP_URL).href;
  const headline = `${wolfName} here: <${suiteHref}|${group.name} tests> ${
    failingRuns.length ? "failed." : "passed!"
  }`;

  const runBlocks = failingRuns.map((run) => {
    return {
      type: "section",
      text: {
        type: "mrkdwn",
        text: `<${environment.APP_URL}/run/${run.id}|${run.test_name}>`,
      },
      accessory: {
        type: "image",
        image_url: run.gif_url,
        alt_text: run.test_name,
      },
    };
  });

  return {
    blocks: [
      {
        type: "context",
        elements: [
          {
            type: "image",
            image_url: `https://storage.googleapis.com/spirit-imgs/wolf-${wolfVariant}-slack.png`,
            alt_text: wolfName,
          },
          {
            type: "mrkdwn",
            text: headline,
          },
        ],
      },
      ...runBlocks,
    ],
  };
};

export const postMessageToSlack = async ({
  message,
  webhook_url,
}: PostMessageToSlack): Promise<void> => {
  const webhook = new IncomingWebhook(webhook_url);
  await webhook.send(message);
};

export const sendSlackAlert = async ({
  group,
  logger,
  runs,
  suite,
}: SendSlackAlert): Promise<void> => {
  const log = logger.prefix("sendSlackAlert");
  log.debug("suite", suite.id);

  try {
    const users = await findUsersForTeam(suite.team_id, { logger });
    const integration = await findIntegration(
      group.alert_integration_id || "",
      {
        logger,
      }
    );

    if (!integration.webhook_url) {
      log.error("no webhook url for integration", integration.id);
      throw new Error("No webhook url for Slack");
    }

    const user = users.length ? (randomChoice(users) as User) : null;

    await postMessageToSlack({
      message: buildMessageForSuite({ group, runs, suite, user }),
      webhook_url: integration.webhook_url,
    });
    log.debug("sent");
  } catch (error) {
    logger.alert("error: Slack alert", error.message);
  }
};
