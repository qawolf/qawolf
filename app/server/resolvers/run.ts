import { db } from "../db";
import environment from "../environment";
import { AuthenticationError } from "../errors";
import { Logger } from "../Logger";
import { findRun, findRunsForSuite, UpdateRun, updateRun } from "../models/run";
import { expireRunner, findRunner } from "../models/runner";
import { postMessageToSlack } from "../services/alert/slack";
import {
  Context,
  InstrumentTestRunMutation,
  ModelOptions,
  Run,
  Suite,
  SuiteRun,
  UpdateRunMutation,
} from "../types";
import { minutesFromNow } from "../utils";
import { ensureSuiteAccess, ensureTestAccess, ensureUser } from "./utils";

type ValidateApiKey = {
  api_key: string | null;
  run: Run;
};

const postRunFailedMessageToSlack = async (
  run: Run,
  logger: Logger
): Promise<void> => {
  if (!environment.SLACK_UPDATES_WEBHOOK || run.status !== "fail") return;

  return postMessageToSlack({
    message: {
      text: `üö® Run failure: run ${run.id} failed (test ${run.test_id})!`,
    },
    webhook_url: environment.SLACK_UPDATES_WEBHOOK,
  }).catch((error) => {
    logger.alert("could not send slack message", error.message);
  });
};

/**
 * @summary Checks that an API key is valid and allows mutating
 *   the given run.
 */
export const validateApiKey = async (
  { api_key, run }: ValidateApiKey,
  { logger, trx }: ModelOptions
): Promise<void> => {
  const log = logger.prefix("validateApiKey");
  log.debug();

  if (!api_key) {
    log.error("no api key provided");
    throw new AuthenticationError("invalid api key");
  }

  const runner = await findRunner({ run_id: run.id }, { logger, trx });

  if (!runner || api_key !== runner.api_key) {
    log.error("incorrect api key for run", run?.id, "runner", runner?.id);
    throw new AuthenticationError("invalid api key");
  }
};

/**
 * @summary Logs the run status and notifies us if a run fails
 */
export const instrumentTestRunResolver = async (
  _: Record<string, unknown>,
  { status, test_id }: InstrumentTestRunMutation,
  { logger, teams, user: contextUser }: Context
): Promise<boolean> => {
  const log = logger.prefix("instrumentTestRun");

  try {
    const user = ensureUser({ logger, user: contextUser });
    await ensureTestAccess({ logger, teams, test_id });

    if (status === "pass") {
      log.debug(`pass (test ${test_id})`);
    } else if (status === "fail") {
      log.debug(`fail (test ${test_id})`);
      if (!environment.SLACK_UPDATES_WEBHOOK) return true;

      await postMessageToSlack({
        message: {
          text: `üïµÔ∏è‚Äç‚ôÄÔ∏è Preview failure: ${user.email} (test ${test_id}, user ${user.id})!`,
        },
        webhook_url: environment.SLACK_UPDATES_WEBHOOK,
      });
    }

    return true;
  } catch (error) {
    log.alert(error.message);

    return false;
  }
};

/**
 * @returns Array of SuiteRuns
 */
export const suiteRunsResolver = async (
  { id }: Suite,
  _: Record<string, unknown>,
  { logger, teams }: Context
): Promise<SuiteRun[]> => {
  logger.debug("suiteRunsResolver");
  await ensureSuiteAccess({ logger, teams, suite_id: id });

  return findRunsForSuite(id, { logger });
};

/**
 * @returns Updated run record
 */
export const updateRunResolver = async (
  _: Record<string, unknown>,
  { current_line, id, status }: UpdateRunMutation,
  { api_key, logger }: Context
): Promise<Run> => {
  const log = logger.prefix("updateRunResolver");
  log.debug(id);

  return db.transaction(async (trx) => {
    const run = await findRun(id, { logger, trx });
    await validateApiKey({ api_key, run }, { logger, trx });

    const updates: UpdateRun = { id };

    if (status === "created") {
      updates.started_at = minutesFromNow();
    } else {
      updates.current_line = current_line;
      updates.status = status;
    }

    if (status === "pass") {
      await expireRunner({ run_id: id }, { logger, trx });
    }

    postRunFailedMessageToSlack(run, logger);

    return updateRun(updates, { logger });
  });
};
