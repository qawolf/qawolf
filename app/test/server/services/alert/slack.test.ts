import { db, dropTestDb, migrateDb } from "../../../../server/db";
import { findRunsForSuite } from "../../../../server/models/run";
import * as slack from "../../../../server/services/alert/slack";
import * as azure from "../../../../server/services/aws/storage";
import { minutesFromNow } from "../../../../server/utils";
import {
  buildGroup,
  buildIntegration,
  buildRun,
  buildSuite,
  buildTeam,
  buildTeamUser,
  buildTest,
  buildUser,
  logger,
} from "../../utils";

const { buildMessageForSuite, sendSlackAlert } = slack;

const group = buildGroup({ alert_integration_id: "integrationId" });
const integration = buildIntegration({});
const user = buildUser({});

beforeAll(async () => {
  await migrateDb();

  await db("users").insert(user);
  await db("teams").insert(buildTeam({}));
  await db("team_users").insert(buildTeamUser({}));

  await db("integrations").insert(integration);
  await db("groups").insert(group);

  await db("suites").insert([buildSuite({})]);
  await db("tests").insert(buildTest({ name: "testName" }));
  return db("runs").insert([
    buildRun({
      completed_at: minutesFromNow(),
      status: "fail",
      suite_id: "suiteId",
    }),
    buildRun({
      completed_at: minutesFromNow(),
      i: 2,
      status: "pass",
      suite_id: "suiteId",
    }),
  ]);
});

afterAll(() => dropTestDb());

describe("buildMessageForSuite", () => {
  beforeAll(() => {
    jest.spyOn(azure, "createStorageReadAccessUrl").mockReturnValue("url");
  });

  afterAll(jest.restoreAllMocks);

  it("builds a Slack message for a failing suite", async () => {
    const group = await db.select("*").from("groups").first();
    const suite = await db.select("*").from("suites").first();
    const runs = await findRunsForSuite(suite.id, { logger });

    expect(
      buildMessageForSuite({ group, runs, suite, user })
    ).toMatchSnapshot();
  });

  it("builds a Slack message for a passing suite", async () => {
    await db("runs").where({ id: "runId" }).update({ status: "pass" });

    const group = await db.select("*").from("groups").first();
    const suite = await db.select("*").from("suites").first();
    const runs = await findRunsForSuite(suite.id, { logger });

    expect(
      buildMessageForSuite({ group, runs, suite, user })
    ).toMatchSnapshot();

    await db("runs").where({ id: "runId" }).update({ status: "fail" });
  });
});

describe("sendSlackAlert", () => {
  beforeAll(() => {
    jest.spyOn(slack, "postMessageToSlack").mockResolvedValue();
  });

  afterEach(jest.clearAllMocks);

  it("sends Slack alert", async () => {
    const group = await db.select("*").from("groups").first();
    const suite = await db.select("*").from("suites").first();
    const runs = await findRunsForSuite(suite.id, { logger });

    await sendSlackAlert({ group, logger, runs, suite });

    expect(slack.postMessageToSlack).toBeCalledWith({
      message: buildMessageForSuite({ group, runs, suite, user }),
      webhook_url: integration.webhook_url,
    });
  });
});
