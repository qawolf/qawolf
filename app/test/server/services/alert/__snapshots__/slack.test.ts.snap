// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildMessageForSuite builds a Slack message for a failing suite 1`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://storage.googleapis.com/spirit-imgs/wolf-brown-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/tests/suiteId|group1 tests> failed.",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
    Object {
      "accessory": Object {
        "alt_text": "testName",
        "image_url": "url",
        "type": "image",
      },
      "text": Object {
        "text": "<http://localhost:3000/run/runId|testName>",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
}
`;

exports[`buildMessageForSuite builds a Slack message for a passing suite 1`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://storage.googleapis.com/spirit-imgs/wolf-brown-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/tests/suiteId|group1 tests> passed!",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
  ],
}
`;
